### ISSUE TITLE
Please figure out hermetic way to include numpy and cv2 packages

### ISSUE THREAD

**Issue by kenrussell** [catapult-project/catapult#2865]
Telemetry's pure-Python PNG reader is the bottleneck in several key GPU tests, including pixel_test and screenshot_sync_test.

Some platforms can easily use "pip install" to install the cv2 (OpenCV) and numpy packages; Linux specifically. However, on macOS at least, this is not possible. The versions in the pip database are obsolete.

@nedn figured out a way to build (if I recall correctly) numpy so that it could be loaded from the build directory rather than installed. I think this was done in these two CLs:

https://codereview.chromium.org/1913083003/
https://codereview.chromium.org/1916613002/

However, this work was never pushed across the finish line.

Please complete this work. It's urgent for Chrome's GPU team in order to speed up the execution of these tests.

Thanks.

**Comment by kenrussell** [comment]
CC @nedn @ehanley324 .

**Comment by nedn** [comment]
This is blocked on @ehanley324 's work to fetch all telemetry dependencies before runtime (idempotency), otherwise, I don't know how to download those file during runtime then importing them during runtime.
Checking this binary to the tree directly is not an option due to their size.

**Comment by kenrussell** [comment]
I see. Thanks for the info. Is there an issue filed for that one so that this could be "blocked" on it (though it doesn't look like Github's issues have that concept)?

**Comment by nedn** [comment]
Blocking issue: https://bugs.chromium.org/p/chromium/issues/detail?id=549140

**Comment by ehanley324** [comment]
Do these binaries currently exist in cloud storage?  In crbug.com/549140 I am adding a script to fetch dependencies from cloud storage to a local path so they can be built in the isolate that is transferred to the bots.  These two dependencies need to be in cloud storage and the json config defining them in a client config that could be passed to this fetch dependency step in the recipe.  

This can be done regardless of the tests being idempotent as when telemetry runs and calls FetchPath for the binary, it will see it locally and won't attempt to download it.

**Comment by kenrussell** [comment]
I'm not sure they do currently exist in cloud storage. Ned recently figured out a way to build these packages so they'd be picked up correctly without needing to be "installed". It would be awesome to finish this work and document the process.

**Comment by nedn** [comment]
To build opencv (cv2), "-D BUILD_SHARED_LIBS=OFF" need to be added. 

To build numpy, just use "pip install numpy" & copy the whole numpy package from installed directory.

**Comment by ehanley324** [comment]
So this will actually have to be part of the script to fetch the dependencies since it happens outside of the dependency manager.  

How are these accessed then?  Where in third_party/catapult should they be downloaded to?

**Comment by nedn** [comment]
Both should be extractable to third_party/ directly. Just please make sure that their content in the zip files have the double folders structure, (cv2/cv2/ & numpy/numpy/)

**Comment by kenrussell** [comment]
Could we please reopen this conversation? The long run time of these tests is becoming a significant problem. It looks like some progress was made on http://crbug.com/549140 ; are we in a position where we could now download the binaries for cv2 and numpy from cloud storage during the build process?

**Comment by ehanley324** [comment]
Unfortunately the idempotency stuff is about 60% done and wasn't high enough priority to finish so it might be on hold for a while.  I will take a look in the next day or two to see if it is something we can do without finishing that work and try and tackle it if it is.

**Comment by kenrussell** [comment]
Could we run a step during "gclient sync" which would download certain Telemetry binaries from cloud storage, so they can be copied into the output directory during the build process? It's OK from my standpoint if we special-case which such binaries are downloaded, instead of solving some larger problem.

**Comment by ehanley324** [comment]
So I briefly looked at the code, the rough plan is as follows: 

1)  Write independent script (probably in telemetry/internal/image_processing) to build numpy and opencv and upload the zip files to some cloud storage hash in the internal telemetry bucket
2) Add entries to the dependency manager json file for these two zip files
3) Replace instances in telemetry that import this module to download the files first if necessary

The open question is on step 3.  I am trying to figure out when I will download and extract the zip files from cloud storage and where I will store them so that external_modules (https://cs.corp.google.com/github/catapult-project/catapult/telemetry/telemetry/internal/util/external_modules.py?q=external+file:%5Ecatapult-project/catapult/telemetry/+package:%5Egithub$&l=28) can import them from the python path (I assume keeping that same external_modules entry point to these two modules is the correct thing to do).

I am not sure if I will be able to add these files to the system path and then import them in the middle of telemetry running, I don't know python that well.  Therefore, I don't know until I try if this will work. Hoping to get to that point this week before I am on leave.

If I can get to work this will be nice for idempotency work since it will be a part of dependency manager and will fit into my pre-fetch script when we get to that point.

**Comment by kenrussell** [comment]
Thanks for investigating this @ehanley324 . If it turns out that it is possible to dynamically modify the Python path in this manner -- specifically, so that the native modules can be picked up -- then I think the best place to store them is in a temporary directory created with Python's tempfile API. src/content/test/gpu/gpu_tests/cloud_storage_integration_test_base.py for example uses this to download reference images from cloud storage on the fly.

**Comment by ehanley324** [comment]
so today is my last day.  I have a CL that outlines the basic solution (https://codereview.chromium.org/2624143004) , but my upload of the zips to cloud storage isn't working correctly in my script, so here are the entries for the binary_dependencies.json once the correct hash is generated.

    "cv2": {
      "cloud_storage_base_folder": "binary_dependencies",
      "cloud_storage_bucket": "chrome-telemetry",
      "file_info": {
        "linux_x86_64": {
          "cloud_storage_hash": "b556bdbf47a9dd746e37dceb999576de5633627c",
          "download_path": "bin/linux/x86_64/cv2.zip"
        }
      }
    },

    "numpy": {
      "cloud_storage_base_folder": "binary_dependencies",
      "cloud_storage_bucket": "chrome-telemetry",
      "file_info": {
        "linux_x86_64": {
          "cloud_storage_hash": "469f2973217654a9ad18da02e6e67b2952b1573b",
          "download_path": "bin/linux/x86_64/numpy.zip"
        }
      }
    },

So to finish this work someone needs to start from this rough CL, flush out the details to get it running and the zips uploaded to cloud storage, update the hashes in binary_dependencies.json and then test it.  

Just for a paper trail of how to build the binaries from the email thread (which we still aren't sure is correct): 

1) numpy
Command: 'pip install numpy'
Output: Requirement already satisfied (use --upgrade to upgrade): numpy in /usr/lib/python2.7/dist-packages
Cleaning up...

I assume I zip up everything in /usr/lib/python2.7/dist-packages/numpy or do I need to do additional build steps before I zip up that file?  

2) opencv
This was a little trickier and more vague.  I tried to follow instructions here:
http://docs.opencv.org/master/d7/d9f/tutorial_linux_install.html

Here are the commands I executed: 
git clone https://github.com/opencv/opencv.git
git clone https://github.com/opencv/opencv_contrib.git
cd opencv
mkdir build
cd build
cmake -D CMAKE_BUILD_TYPE=Release -D BUILD_SHARED_LIBS=OFF -D CMAKE_INSTALL_PREFIX=/usr/local/google/home/eyaich/projects/chromium/src/third_party/catapult/opencv/build /usr/local/google/home/eyaich/projects/chromium/src/third_party/catapult/opencv
make -j7
sudo make install
NOTE: I didn't do anything with step 4 building python, not sure if I should have done an additional step or not.  

So now things don't exist under bin like I thought they would, but under lib and there are two sets of files, neither of which match what is in Ned's CL (https://codereview.chromium.org/1913083003/). 

See the set in /usr/local/google/home/eyaich/projects/chromium/src/third_party/catapult/opencv/build/lib
https://paste.googleplex.com/6753840116269056

and then after the 'sudo make install' I got output saying things like: 
-- Set runtime path of "/usr/local/lib/python2.7/dist-packages/cv2.so" to "/usr/local/lib"
-- Set runtime path of "/usr/local/bin/opencv_traincascade" to "/usr/local/lib"
-- Installing: /usr/local/bin/opencv_createsamples
-- Set runtime path of "/usr/local/bin/opencv_createsamples" to "/usr/local/lib"
-- Installing: /usr/local/bin/opencv_annotation
-- Set runtime path of "/usr/local/bin/opencv_annotation" to "/usr/local/lib"
-- Installing: /usr/local/bin/opencv_visualisation
-- Set runtime path of "/usr/local/bin/opencv_visualisation" to "/usr/local/lib"
-- Installing: /usr/local/bin/opencv_version
-- Set runtime path of "/usr/local/bin/opencv_version" to "/usr/local/lib"

so there is another set of .a files in /usr/local/lib, see that set here: 
https://paste.googleplex.com/5522998387802112

**Comment by kainino0x** [comment]
@kenrussell has asked if I'm interested in working on this. I would be glad to. I plan to work on it pretty soon.

**Comment by perezju** [comment]
Hmm, are we sure that building these packages on some random machine and pushing binaries to cloud storage is the way to go?

I'm having troubles in #3580 because the binaries installed are probably conflicting with other things in my workstation and prevent me from running any Telemetry command at all.

Making the check in [external_modules.py](https://github.com/catapult-project/catapult/blob/08a6e0ac161db7309d8f9cad0ccd38e0b1fd41e0/telemetry/telemetry/internal/util/external_modules.py#L54) non-fatal and adding a bit of extra logging I see:

```
$ tools/perf/run_benchmark --browser list
WARNING:root:Unable to import numpy: /usr/local/google/code/clankium/src/third_party/catapult/telemetry/third_party/numpy/lib/numpy_linux_x86_64_58b6664311a625727f246e3783571dad940bd934/numpy/random/mtrand.so: undefined symbol: PyFPE_jbuf
WARNING:root:Unable to import numpy: cannot import name add_newdocs
WARNING:root:Unable to import psutil: No module named psutil
```

The second attempt to import numpy is what causes the command to bail out without my changes.

**Comment by randomascii** [comment]
Perhaps related to (newly filed) issue 4038?
